src/Analyzer/Resolve/QueryAnalyzer.cpp

```diff
@@ -3102,10 +3145,127 @@ ProjectionNames QueryAnalyzer::resolveFunction(QueryTreeNodePtr & node, Identifi
                     in_second_argument = in_second_argument->cloneAndReplace(table_expression, std::move(replacement_table_expression_table_node));
                 }
             }
+            /// If it's a function node like array(..) or tuple(..), consider rewriting them to 'has':
+            if (auto * non_const_set_candidate = in_second_argument->as<FunctionNode>())
+            {
+                const auto & candidate_name = non_const_set_candidate->getFunctionName();
+
+                /// Case 1: array(..) node – keep existing behavior
+                if (candidate_name == "array")
+                {
+                    bool contains_nullable = false;
+                    for (const auto & array_elem : non_const_set_candidate->getArguments().getNodes())
+                    {
+                        if (isNullableOrLowCardinalityNullable(array_elem->getResultType()))
+                        {
+                            contains_nullable = true;
+                            break;
+                        }
+                    }
+                    if (!contains_nullable)
+                    {
+                        function_name = "has";
+                        is_special_function_in = false;
+                        auto & fn_args = function_node.getArguments().getNodes();
+                        std::swap(fn_args[0], fn_args[1]);
+                    }
+                }
+                /// Case 2: tuple(..) node – rewrite it into an array
+                else if (candidate_name == "tuple")
+                {
+                    bool left_is_null = false;
+                    if (const auto * const_node = in_first_argument->as<ConstantNode>())
+                        left_is_null = const_node->getValue().isNull();
+                    /// Infer a common type among the tuple elements
+                    DataTypePtr common_type = nullptr;
+                    /// Use the type from the left-hand side of the IN statement if it's not NULL
+                    if (!left_is_null)
+                    {
+                        common_type = in_first_argument->getResultType();
+                    }
+                    else
+                    {
+                        auto & tuple_args = non_const_set_candidate->getArguments().getNodes();
+                        for (const auto & arg : tuple_args)
+                        {
+                            if (!common_type)
+                                common_type = arg->getResultType(); // start with the first element’s type
+                            else
+                                common_type = tryGetLeastSupertype(DataTypes{common_type, arg->getResultType()});
+                        }
+                    }
```

Comment:
Let's do the following. If left arg is Null, and there is any nullable type, rewrite it to
```
arrayCount([isNull(arg1), isNull(arg2), ...]) > 0
```
for all nullable args.
(or to constant 0, if no nullable args)